### ISSUE TITLE
Difficult to initialise cf i18n from plugin (choice between vendored dependency clash, null dereference, or hardcoding to English?)

### ISSUE THREAD

**Issue by holly-cummins** [cloudfoundry/cli#1018]
The [cf-blue-green-deploy plugin](https://github.com/bluemixgaragelondon/cf-blue-green-deploy) is having some challenges with the cf cli i18n. 

If we build the plugin with Go 1.6+, we get the following error, caused by the behaviour change with vendored dependencies in Go 1.6:

    ./main.go:173: cannot assign "github.com/nicksnyder/go-i18n/i18n".TranslateFunc to "github.com/cloudfoundry/cli/cf/i18n".T (type "github.com/cloudfoundry/cli/vendor/github.com/nicksnyder/go-i18n/i18n".TranslateFunc) in multiple assignment

The exact line of code that fails is:

    // T needs to point to a translate func, otherwise cf internals blow up
    i18n.T, _ = go_i18n.Tfunc("")

The discussion on https://github.com/cloudfoundry/cli/issues/889 about this issue says that setting i18n.T is not supported, which is fine, since we don't actually want to set T anyway. The suggested fix on #889   is just to remove the assignment of i18n.T. If we do that, everything compiles fine, but at runtime we see the following:

    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6c2da] 

on the first thing in `manifest.go` which tries to do a translation:

 sliceErr := fmt.Errorf(T("Expected {{.PropertyName}} to be a list of strings.", map[string]interface{}{"PropertyName": key}))

We could make things work by forcing the language to English in our plugin, by doing something like 

    +type localeGetter struct{}
    +
    +func (l localeGetter) Locale() string {
    +       return "en-us"
    +}
    +
     var _ = Describe("Manifest reader", func() {
        // testing code that calls into cf cli requires T to point to a translate func
     -       i18n.T, _ = go_i18n.Tfunc("")
     +       i18n.T = i18n.Init(localeGetter{})
     +type localeGetter struct{}
    +
    +func (l localeGetter) Locale() string {
    +       return "en-us"
    +}

However, forcing the language to English in order to avoid null references in the cf cli doesn't seem right. Is there some initialisation of the cf cli that we're missing? Are we going down a code path which somehow bypasses the normal initialisation?

**Comment by XenoPhex** [comment]
Howdy @holly-cummins,

I'm kind of at an impasse; I want to take the time to help figure out how to fix this issue, but at the same time, I want to stop you from doing something you shouldn't. The CF CLI team's official policy on all packages outside of the `plugin` directory is the following: Do not import this package externally. None of the code in these packages were ever designed/intended to be used outside of the CF CLI (including for plugins).

Case in point, you're now facing an issue in regards to the internationalization code that was designed to (only) work inside the CF CLI codebase. So here's my recommendation on fixing your issue: 
- Write your own manifest parser 
- Copy ours into your codebase and remove the i18n code (thus making it yours to maintain).

To be honest, the code in `manifest.go` is a freakin' mess, and you probably don't need 95% of it (and frankly I don't trust most of it either). You're *very much better off* writing your own.

**Comment by holly-cummins** [comment]
Thanks, @XenoPhex, this is exactly what we needed to know. We didn't know about the internal/external package convention within the cf cli (and in fact when we started working on the our plugin I'm not sure the plugin API even existed). Sticking to externals definitely seems like the right thing to do, and hopefully will fix another problem we have, which is that we're very sensitive to upgrades in the cf cli. 
We've started looking at a combination of your two suggested approaches. We're probably not the only plugin which needs to read manifest contents, so I'll take the opportunity to put in a plug for including manifest parsing in the plugin API at some point. :)
